src/internal/pjsdb/job_crud.go

```diff
@@ -307,17 +307,28 @@ func DeleteJob(ctx context.Context, tx *pachsql.Tx, id JobID) ([]JobID, error) {
 // done timestamp in database.
 func ErrorJob(ctx context.Context, tx *pachsql.Tx, jobID JobID, errCode pjs.JobErrorCode) error {
 	ctx = pctx.Child(ctx, "complete error")
+	errStr := errEnumToString(errCode)
 	_, err := tx.ExecContext(ctx, `
 		UPDATE pjs.jobs
 		SET done = CURRENT_TIMESTAMP, error = $1
 		WHERE id = $2 AND error IS NULL
-	`, errCode, jobID)
+	`, errStr, jobID)
 	if err != nil {
 		return errors.Wrapf(err, "error job: update error and state to done")
 	}
 	return nil
 }
 
+// errEnumToString fills in the gap between pjs proto error enum and database error enum
+func errEnumToString(errCode pjs.JobErrorCode) string {
+	converter := map[pjs.JobErrorCode]string{
+		pjs.JobErrorCode_DISCONNECTED: "disconnected",
+		pjs.JobErrorCode_FAILED:       "failed",
+		pjs.JobErrorCode_CANCELED:     "cancelled",
+	}
+	return converter[errCode]
+}
+
```

Comment:
It’s surprising that `pjsdb` converts a `pjs.JobErrorCode` to a PostgreSQL enum string, but does not convert the enum string abck to a Job Error Code enum (that seems to happen in the internal PJS server package.

I believe that `pjsdb` should completely abstract over PostgreSQL and use the same types for input and output.

Reply:
Hey Bob, thank you the comment and review. I read through this comment several times, but I'm not sure what your suggested change is. Could you elaborate on that? Thanks.

Reply:
Sure thing! What I mean is that right now `pjsdb.ErrorJob` is called by the internal `pjs` package with a proto `pjs.JobErrorCode` enum; your code here converts that error code enum into a PostgreSQL enum string and uses it internally. However, `pjsdb.GetJob` returns a `pjsdb.Job` rather than a proto `pjs.Job`, which means that the internal `pjs` package must know how to convert a `pjsdb.Job` into a `pjs.JobInfo` using the unexported `toJobInfo`. That function uses the proto `pjs` package’s `pjs.JobErrorCode_value` map from strings to int32 enums.

So one package knows how to turn enums into strings, while another package knows how to use a _third_ package to turn strings into enums!

There’s an even bigger problem here: the three packages don’t agree! The PostgreSQL enum that `pjsdb` knows about comprises the three-member set (`disconnected`, `failed`, `cancelled`), while the _four-member_ set (`JobErrorCode_UNSPECIFIED`, `FAILED`, `DISCONNECTED`, `CANCELED`). Note that there are differences in both capitalization and spelling (`cancelled` is international English, whiled `CANCELED` is American English), as well as in cardinality!

What I think you should do is encapsulate all the knowledge of how to convert the proto messages into PostgreSQL tables, and how to convert PostgreSQL tables into proto messages, inside of `pjsdb` and have the internal `pjs` package just send those protos the `pjsdb` and receive them back.